File: app/src/test/java/com/arv/app/core/data/InvitationTest.kt
package com.arv.app.core.data

import com.arv.app.core.data.local.InviteEntity
import com.arv.app.core.data.local.MemberEntity
import com.arv.app.core.model.MemberRole
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class InvitationTest {

    private val code = "K7M2QX"

    private fun invite(
        usedAt: Long? = null,
        usedBy: String? = null,
        revokedAt: Long? = null,
        issuer: String = "u_ruth",
        role: MemberRole = MemberRole.CONTRIBUTOR
    ) = InviteEntity(
        code = code,
        familyId = "fam_1",
        issuedByUserId = issuer,
        grantsRole = role,
        createdAt = 1_000L,
        usedAt = usedAt,
        usedByUserId = usedBy,
        revokedAt = revokedAt
    )

    @Test
    fun `a good code writes the inviter into the member row`() {
        val r = Invitation.redeem("k7m-2qx", invite(), null, "u_dana", 5_000L)
        assertTrue(r is Invitation.Result.Accepted)
        val accepted = r as Invitation.Result.Accepted
        assertEquals("u_ruth", accepted.member.invitedBy)
        assertEquals("u_dana", accepted.member.userId)
        assertEquals("fam_1", accepted.member.familyId)
        assertEquals(MemberRole.CONTRIBUTOR, accepted.member.role)
        assertEquals(5_000L, accepted.member.joinedAt)
    }

    @Test
    fun `accepting spends the code and records who spent it`() {
        val r = Invitation.redeem(code, invite(), null, "u_dana", 5_000L)
        val spent = (r as Invitation.Result.Accepted).spent
        assertEquals(5_000L, spent.usedAt)
        assertEquals("u_dana", spent.usedByUserId)
    }

    @Test
    fun `a spent code says so instead of pretending it never existed`() {
        val r = Invitation.redeem(code, invite(usedAt = 2_000L, usedBy = "u_kev"), null, "u_dana", 5_000L)
        assertEquals(Invitation.Result.AlreadyUsed, r)
    }

    @Test
    fun `a forwarded code cannot admit a second person under the same inviter`() {
        val first = Invitation.redeem(code, invite(), null, "u_dana", 5_000L)
        val spent = (first as Invitation.Result.Accepted).spent
        val second = Invitation.redeem(code, spent, null, "u_someone_else", 6_000L)
        assertEquals(Invitation.Result.AlreadyUsed, second)
    }

    @Test
    fun `revoked is not the same answer as used`() {
        val r = Invitation.redeem(code, invite(revokedAt = 3_000L), null, "u_dana", 5_000L)
        assertEquals(Invitation.Result.Revoked, r)
    }

    @Test
    fun `nobody invites themselves`() {
        val r = Invitation.redeem(code, invite(issuer = "u_dana"), null, "u_dana", 5_000L)
        assertEquals(Invitation.Result.YourOwn, r)
    }

    @Test
    fun `already a member is a no-op, not a failure`() {
        val existing = MemberEntity(
            familyId = "fam_1", userId = "u_dana",
            role = MemberRole.KEEPER, joinedAt = 100L
        )
        val r = Invitation.redeem(code, invite(), existing, "u_dana", 5_000L)
        assertEquals(Invitation.Result.AlreadyInThisFamily, r)
    }

    @Test
    fun `half typed is not a code and not an error`() {
        assertEquals(Invitation.Result.NotACode, Invitation.redeem("K7M", invite(), null, "u_dana", 5_000L))
        assertEquals(Invitation.Result.NotACode, Invitation.redeem(null, invite(), null, "u_dana", 5_000L))
    }

    @Test
    fun `well formed but unknown is its own answer`() {
        assertEquals(Invitation.Result.Unknown, Invitation.redeem("ABCDEF", null, null, "u_dana", 5_000L))
    }

    @Test
    fun `the role travels with the invitation, not with the joiner`() {
        val r = Invitation.redeem(code, invite(role = MemberRole.VIEWER), null, "u_dana", 5_000L)
        assertEquals(MemberRole.VIEWER, (r as Invitation.Result.Accepted).member.role)
        assertNull(r.member.personId)
    }
}

File: functions/test/invitation.test.js
// Mirrors app/src/test/java/com/arv/app/core/data/InvitationTest.kt. Same fixtures,
// same cases, same expected answers. Run with `npm test` in this folder; needs nothing
// installed beyond Node.

const test = require("node:test");
const assert = require("node:assert/strict");
const { normalize, decide } = require("../invitation");

const invite = (over = {}) => ({
  code: "K7M2QX",
  familyId: "fam_1",
  createdBy: "u_ruth",
  role: "CONTRIBUTOR",
  createdAt: 1000,
  expiresAt: null,
  usedAt: null,
  usedBy: null,
  revokedAt: null,
  familyName: "Delaney",
  ...over
});

test("normalize accepts every way a code gets written down", () => {
  assert.equal(normalize("k7m-2qx"), "K7M2QX");
  assert.equal(normalize("K7M2QX"), "K7M2QX");
  assert.equal(normalize("k 7 m 2 q x"), "K7M2QX");
});

test("normalize refuses what could never be a code", () => {
  assert.equal(normalize("K7M"), null);
  assert.equal(normalize("K7M2QXA"), null);
  assert.equal(normalize("K7M2Q0"), null, "0 is not in the alphabet");
  assert.equal(normalize("K7M2QI"), null, "nor is I");
  assert.equal(normalize(null), null);
  assert.equal(normalize(undefined), null);
  assert.equal(normalize(42), null);
});

test("a good code is accepted", () => {
  assert.deepEqual(decide(invite(), false, "u_dana", 5000), { status: "ACCEPTED" });
});

test("no such code", () => {
  assert.deepEqual(decide(null, false, "u_dana", 5000), { status: "UNKNOWN" });
});

test("a spent code says so", () => {
  assert.deepEqual(decide(invite({ usedAt: 2000, usedBy: "u_kev" }), false, "u_dana", 5000), { status: "ALREADY_USED" });
});

test("a withdrawn code says so", () => {
  assert.deepEqual(decide(invite({ revokedAt: 3000 }), false, "u_dana", 5000), { status: "REVOKED" });
});

test("a code past its end is expired", () => {
  assert.deepEqual(decide(invite({ expiresAt: 4000 }), false, "u_dana", 5000), { status: "EXPIRED" });
});

test("a code is expired the instant it reaches its end", () => {
  assert.deepEqual(decide(invite({ expiresAt: 5000 }), false, "u_dana", 5000), { status: "EXPIRED" });
});

test("a code minted before codes had an end never runs out", () => {
  assert.deepEqual(decide(invite({ expiresAt: null }), false, "u_dana", Number.MAX_SAFE_INTEGER), { status: "ACCEPTED" });
});

test("being spent is a better answer than having run out", () => {
  assert.deepEqual(decide(invite({ usedAt: 2000, expiresAt: 3000 }), false, "u_dana", 5000), { status: "ALREADY_USED" });
});

test("being withdrawn is a better answer than having run out", () => {
  assert.deepEqual(decide(invite({ revokedAt: 2000, expiresAt: 3000 }), false, "u_dana", 5000), { status: "REVOKED" });
});

test("nobody invites themselves", () => {
  assert.deepEqual(decide(invite({ createdBy: "u_dana" }), false, "u_dana", 5000), { status: "YOUR_OWN" });
});

test("already in the family is not an error, just nothing to do", () => {
  assert.deepEqual(decide(invite(), true, "u_dana", 5000), { status: "ALREADY_IN_THIS_FAMILY" });
});
